qa/L0_metrics/pinned_memory_metrics_test.py

```diff
@@ -0,0 +1,177 @@
+    r.raise_for_status()
+
+    total_bytes_match = total_bytes_pattern.search(r.text)
+    total_bytes_value = total_bytes_match.group(1)
+
+    used_bytes_match = used_bytes_pattern.search(r.text)
+    used_bytes_value = used_bytes_match.group(1)
+
+    return total_bytes_value, used_bytes_value
+
+
+class TestPinnedMemoryMetrics(unittest.TestCase):
+    def setUp(self):
+        self.inference_completed = threading.Event()
+
+        shape = [1, 16]
+        self.model_name = "libtorch_float32_float32_float32"
+        input0_data = np.random.rand(*shape).astype(np.float32)
+        input1_data = np.random.rand(*shape).astype(np.float32)
+
+        self.inputs = [
+            httpclient.InferInput(
+                "INPUT0", input0_data.shape, "FP32"
+            ).set_data_from_numpy(input0_data),
+            httpclient.InferInput(
+                "INPUT1", input1_data.shape, "FP32"
+            ).set_data_from_numpy(input1_data),
+        ]
+
+        self.outputs = [
+            httpclient.InferRequestedOutput("OUTPUT__0"),
+            httpclient.InferRequestedOutput("OUTPUT__1"),
+        ]
+
+        # Before loading the model
+        self._assert_pinned_memory_utilization()
+
+    def _assert_pinned_memory_utilization(self):
+        total_bytes_value, used_bytes_value = _get_metrics()
+        self.assertEqual(int(total_bytes_value), TOTAL_PINNED_MEMORY_SIZE)
+        self.assertEqual(int(used_bytes_value), DEFAULT_USED_PINNED_MEMORY_SIZE)
+
+    def _collect_metrics(self):
+        while not self.inference_completed.is_set():
+            total_bytes_value, used_bytes_value = _get_metrics()
+            self.assertEqual(int(total_bytes_value), TOTAL_PINNED_MEMORY_SIZE)
+            # Assert pinned memory usage is within anticipated values
+            self.assertIn(int(used_bytes_value), [0, 64, 128, 192, 256])
+
+    def test_pinned_memory_metrics_asynchronous_requests(self):
+        with httpclient.InferenceServerClient(
+            url="localhost:8000", concurrency=10
+        ) as client:
+            if not client.is_model_ready(self.model_name):
+                client.load_model(self.model_name)
+
+            # Before starting the inference
+            self._assert_pinned_memory_utilization()
+
+            # Start a thread to collect metrics asynchronously
+            metrics_thread = threading.Thread(target=self._collect_metrics)
+            metrics_thread.start()
+
+            # Asynchronous inference requests
+            async_requests = []
+            for _ in range(100):
+                async_requests.append(
+                    client.async_infer(
+                        model_name=self.model_name,
+                        inputs=self.inputs,
+                        outputs=self.outputs,
+                    )
+                )
+
+            time.sleep(1)
+
+            # Set the event to indicate that inference is completed
+            self.inference_completed.set()
```

Comment:
Inference isn't complete in async case until all the results are finished, so I think this should be moved after all the `get_result()` have finished.